[DCOS-37660] Add initial diagnostics script.

mpereira opened the pull request:
**May 25:**
Initial WIP pushed:

**Jul 30:**
With the commits I just pushed I believe this is now a Good Enough™ diagnostics tool for SDK services.

What is done:

-   [X] Fetching of DC/OS version, Mesos tasks, Marathon apps, service configuration, pod status, plans statuses (deploy, recovery, backup-s3, etc.)
-   [X] Fetching of scheduler stderr/stdout logs
-   [X] Fetching of task stderr/stdout logs
-   [X] Fetching of task executor stderr/stdout logs
-   [X] Fetching of killed/failed task stderr/stdout logs (in theory, if Mesos hasn't GC'd the sandbox, the script should be able to do this)
-   [X] Fetching of base-tech specific diagnostics commands (Cassandra "nodetool status", Elastic "node stats", etc.)
-   [X] An easy way to extend base-tech specific diagnostics commands
-   [X] Retrying on all operations where it makes sense
-   [X] Temporary Docker image pushed to [mpereira/dcos-commons:diagnostics](https://hub.docker.com/r/mpereira/dcos-commons/)

I re-used some logic from sdk_diag for the fetching of log files. Some of the functions there were a bit too complected with tests (their original intent), so I couldn't use them verbatim. I believe some of the functions in the "create bundle" script could be moved to modules. I didn't really worry about that.

Possible future work:

-   [ ] Concatenate log files (e.g. stdout+stdout.1+stdout.2 -> stdout)
-   [ ] Create archive from output directory
-   [ ] Add interactive inputs and helpful output to make sure user is running diagnostics against the right cluster/service

Still want to do before merging the PR:

-   [ ] Provide command so that people can run something like (maybe provide a wrapper script?):

    ```bash
    docker run --rm mpereira/dcos-commons:diagnostics bash -c 'PYTHONPATH=/dcos-commons-dist/testing python3 /dcos-commons-dist/tools/create_service_bundle.py hdfs /data-services/hdfs'
    ```
-   [ ] Add docstrings to functions
-   [ ] Write usage documention
-   [ ] Use argparse

Example run:

```text
$ PYTHONPATH=testing ./tools/create_service_bundle.py hdfs /data-services/hdfs
...
(a bunch of output)
...
$ tree hdfs_data-services__hdfs_20180730222531
hdfs_data-services__hdfs_20180730222531
├── dcos_marathon_task_list.json
├── dcos_task.json
├── dcos_version.txt
├── service_configuration.json
├── service_plan_status_deploy.json
├── service_plan_status_recovery.json
├── service_plan_status_replace.json
├── service_pod_status.json
└── tasks
    ├── data-services.hdfs__data-0-node__de194f84-be32-485c-9cd4-f484752b3c78
    │   ├── executor
    │   │   ├── stderr
    │   │   ├── stderr.1
    │   │   ├── stderr.2
    │   │   ├── stderr.3
    │   │   ├── stderr.4
    │   │   ├── stderr.5
    │   │   ├── stderr.6
    │   │   ├── stderr.7
    │   │   ├── stderr.8
    │   │   ├── stderr.9
    │   │   └── stdout
    │   └── task
    │       ├── stderr
    │       ├── stderr.1
    │       ├── stderr.2
    │       ├── stderr.3
    │       ├── stderr.4
    │       ├── stderr.5
    │       ├── stderr.6
    │       ├── stderr.7
    │       ├── stderr.8
    │       ├── stderr.9
    │       └── stdout
    ├── data-services.hdfs__data-1-node__7ed21542-0d9a-4ed3-b722-1418502329af
    │   ├── executor
    │   │   ├── stderr
    │   │   ├── stderr.1
    │   │   ├── stderr.2
    │   │   ├── stderr.3
    │   │   ├── stderr.4
    │   │   ├── stderr.5
    │   │   ├── stderr.6
    │   │   ├── stderr.7
    │   │   ├── stderr.8
    │   │   ├── stderr.9
    │   │   └── stdout
    │   └── task
    │       ├── stderr
    │       ├── stderr.1
    │       ├── stderr.2
    │       ├── stderr.3
    │       ├── stderr.4
    │       ├── stderr.5
    │       ├── stderr.6
    │       ├── stderr.7
    │       ├── stderr.8
    │       ├── stderr.9
    │       └── stdout
    ├── data-services.hdfs__data-2-node__d92d7c06-287e-4f8a-abc0-e21df8a1f2d5
    │   ├── executor
    │   │   ├── stderr
    │   │   ├── stderr.1
    │   │   ├── stderr.2
    │   │   ├── stderr.3
    │   │   ├── stderr.4
    │   │   ├── stderr.5
    │   │   ├── stderr.6
    │   │   ├── stderr.7
    │   │   ├── stderr.8
    │   │   ├── stderr.9
    │   │   └── stdout
    │   └── task
    │       ├── stderr
    │       ├── stderr.1
    │       ├── stderr.2
    │       ├── stderr.3
    │       ├── stderr.4
    │       ├── stderr.5
    │       ├── stderr.6
    │       ├── stderr.7
    │       ├── stderr.8
    │       ├── stderr.9
    │       └── stdout
    ├── data-services.hdfs__journal-0-node__0cdab3d7-4710-4af7-adfe-46e49c07e60d
    ├── data-services.hdfs__journal-1-node__45c970b0-d91b-4c98-a7d6-71832acdea42
    ├── data-services.hdfs__journal-1-node__ef679bae-6e04-45bd-a68e-9b655c9992a1
    ├── data-services.hdfs__journal-2-node__5c860fd8-0e89-4a14-9f8f-f40cc1f835aa
    ├── data-services.hdfs__name-0-node__370fa32f-ea99-443e-8e52-720cb3d9d67c
    ├── data-services.hdfs__name-0-zkfc__461ccd17-dbfd-439d-bd7e-0f55ffadb096
    ├── data-services_hdfs.942d286b-90c3-11e8-bb4f-4aed11d552b2
    ├── journal-0-node__95cd3ae4-7127-48ee-b358-965895706454
    │   ├── executor
    │   │   ├── stderr
    │   │   └── stdout
    │   └── task
    │       ├── stderr
    │       ├── stderr.1
    │       ├── stderr.2
    │       └── stdout
    ├── journal-1-node__3eb48744-24a7-48d5-a619-4f8ffb95dd8a
    │   ├── executor
    │   │   ├── stderr
    │   │   └── stdout
    │   └── task
    │       ├── stderr
    │       └── stdout
    ├── journal-2-node__d5c3ecef-6fb0-41df-a8b0-a24e8fad7388
    │   ├── executor
    │   │   ├── stderr
    │   │   └── stdout
    │   └── task
    │       ├── stderr
    │       ├── stderr.1
    │       ├── stderr.2
    │       └── stdout
    ├── name-0-node__4dabc9fc-2d29-4dae-9674-a9f83f20264d
    │   ├── executor
    │   │   ├── stderr
    │   │   ├── stderr.1
    │   │   └── stdout
    │   └── task
    │       ├── stderr
    │       ├── stderr.1
    │       ├── stderr.2
    │       ├── stderr.3
    │       ├── stderr.4
    │       ├── stderr.5
    │       ├── stderr.6
    │       ├── stderr.7
    │       ├── stderr.8
    │       └── stdout
    └── name-0-zkfc__e5def1db-75f0-466b-9131-e107784f25b5
        ├── executor
        │   ├── stderr
        │   └── stdout
        └── task
            ├── stderr
            ├── stderr.1
            ├── stderr.2
            ├── stderr.3
            ├── stderr.4
            ├── stderr.5
            ├── stderr.6
            ├── stderr.7
            ├── stderr.8
            └── stdout

32 directories, 115 files
```

mpereira commented:
Example run of current WIP:

```
$ PYTHONPATH=testing ./tools/create_service_bundle.py cassandra cassandra
Creating directory cassandra_cassandra_2018052539
(CLI) dcos task --json
Writing file cassandra_cassandra_2018052539/dcos_task.json
(CLI) dcos marathon task list --json
Writing file cassandra_cassandra_2018052539/dcos_marathon_task_list.json
(CLI) dcos cassandra --name=cassandra describe
Writing file cassandra_cassandra_2018052539/service_configuration.json
(CLI) dcos cassandra --name=cassandra pod status --json
Writing file cassandra_cassandra_2018052539/service_pod_status.json
(CLI) dcos cassandra --name=cassandra plan list
(CLI) dcos cassandra --name=cassandra plan status deploy --json
Writing file cassandra_cassandra_2018052539/service_plan_status_deploy.json
(CLI) dcos cassandra --name=cassandra plan status recovery --json
Writing file cassandra_cassandra_2018052539/service_plan_status_recovery.json
(CLI) dcos cassandra --name=cassandra plan status repair --json
Writing file cassandra_cassandra_2018052539/service_plan_status_repair.json
(CLI) dcos cassandra --name=cassandra plan status restore-s3 --json
Writing file cassandra_cassandra_2018052539/service_plan_status_restore-s3.json
(CLI) dcos cassandra --name=cassandra plan status backup-s3 --json
Writing file cassandra_cassandra_2018052539/service_plan_status_backup-s3.json
(CLI) dcos cassandra --name=cassandra plan status backup-azure --json
Writing file cassandra_cassandra_2018052539/service_plan_status_backup-azure.json
(CLI) dcos cassandra --name=cassandra plan status restore-azure --json
Writing file cassandra_cassandra_2018052539/service_plan_status_restore-azure.json
(CLI) dcos cassandra --name=cassandra plan status cleanup --json
Writing file cassandra_cassandra_2018052539/service_plan_status_cleanup.json
(CLI) dcos cassandra --name=cassandra plan status replace --json
Writing file cassandra_cassandra_2018052539/service_plan_status_replace.json
Creating Cassandra bundle
Sending HTTP ['get'] to ['https://mpereira-elasticl-1hniohuxvq79r-1730082276.eu-central-1.elb.amazonaws.com/mesos/master/state.json']: {'Accept': 'application/json'}
Received HTTP response [200]: {'Server': 'openresty', 'Date': 'Fri, 25 May 2018 21:45:54 GMT', 'Content-Type': 'application/json', 'Content-Length': '69025', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip'}
(CLI) dcos task exec node-2-server__c7a678e7-dbcf-4fab-a9e1-97667a57fcd7 bash -c 'export JAVA_HOME=$(ls -d ${MESOS_SANDBOX}/jdk*/jre/) && export TASK_IP=$(${MESOS_SANDBOX}/bootstrap --get-task-ip) && CASSANDRA_DIRECTORY=$(ls -d ${MESOS_SANDBOX}/apache-cassandra-*/) && ${CASSANDRA_DIRECTORY}/bin/nodetool status'
STDOUT:
Datacenter: datacenter1
=======================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address     Load       Tokens       Owns (effective)  Host ID                               Rack
UN  10.0.1.80   84.8 KB    256          100.0%            8e13bb73-2ef8-4e27-9a38-32e091d3894a  rack1
UN  10.0.2.134  69.37 KB   256          100.0%            d94a3a07-b3ab-433c-92d8-50df01532212  rack1
UN  10.0.0.124  88.94 KB   256          100.0%            920265f4-07b6-4b4d-a109-891b3cc10c03  rack1
Writing file cassandra_cassandra_2018052539/cassandra_nodetool_status_node-2-server__c7a678e7-dbcf-4fab-a9e1-97667a57fcd7.txt
(CLI) dcos task exec node-1-server__9a04a418-3e46-429b-aafe-9003ab1bf357 bash -c 'export JAVA_HOME=$(ls -d ${MESOS_SANDBOX}/jdk*/jre/) && export TASK_IP=$(${MESOS_SANDBOX}/bootstrap --get-task-ip) && CASSANDRA_DIRECTORY=$(ls -d ${MESOS_SANDBOX}/apache-cassandra-*/) && ${CASSANDRA_DIRECTORY}/bin/nodetool status'
STDOUT:
Datacenter: datacenter1
=======================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address     Load       Tokens       Owns (effective)  Host ID                               Rack
UN  10.0.1.80   84.8 KB    256          100.0%            8e13bb73-2ef8-4e27-9a38-32e091d3894a  rack1
UN  10.0.2.134  69.37 KB   256          100.0%            d94a3a07-b3ab-433c-92d8-50df01532212  rack1
UN  10.0.0.124  88.94 KB   256          100.0%            920265f4-07b6-4b4d-a109-891b3cc10c03  rack1
Writing file cassandra_cassandra_2018052539/cassandra_nodetool_status_node-1-server__9a04a418-3e46-429b-aafe-9003ab1bf357.txt
(CLI) dcos task exec node-0-server__c98911be-c2ae-4121-aa78-60a6840bd886 bash -c 'export JAVA_HOME=$(ls -d ${MESOS_SANDBOX}/jdk*/jre/) && export TASK_IP=$(${MESOS_SANDBOX}/bootstrap --get-task-ip) && CASSANDRA_DIRECTORY=$(ls -d ${MESOS_SANDBOX}/apache-cassandra-*/) && ${CASSANDRA_DIRECTORY}/bin/nodetool status'
STDOUT:
Datacenter: datacenter1
=======================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address     Load       Tokens       Owns (effective)  Host ID                               Rack
UN  10.0.1.80   84.8 KB    256          100.0%            8e13bb73-2ef8-4e27-9a38-32e091d3894a  rack1
UN  10.0.2.134  69.37 KB   256          100.0%            d94a3a07-b3ab-433c-92d8-50df01532212  rack1
UN  10.0.0.124  88.94 KB   256          100.0%            920265f4-07b6-4b4d-a109-891b3cc10c03  rack1
Writing file cassandra_cassandra_2018052539/cassandra_nodetool_status_node-0-server__c98911be-c2ae-4121-aa78-60a6840bd886.txt
Sending HTTP ['get'] to ['https://mpereira-elasticl-1hniohuxvq79r-1730082276.eu-central-1.elb.amazonaws.com/mesos/master/state.json']: {'Accept': 'application/json'}
Received HTTP response [200]: {'Server': 'openresty', 'Date': 'Fri, 25 May 2018 21:46:23 GMT', 'Content-Type': 'application/json', 'Content-Length': '69025', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip'}
(CLI) dcos task exec node-2-server__c7a678e7-dbcf-4fab-a9e1-97667a57fcd7 bash -c 'export JAVA_HOME=$(ls -d ${MESOS_SANDBOX}/jdk*/jre/) && export TASK_IP=$(${MESOS_SANDBOX}/bootstrap --get-task-ip) && CASSANDRA_DIRECTORY=$(ls -d ${MESOS_SANDBOX}/apache-cassandra-*/) && ${CASSANDRA_DIRECTORY}/bin/nodetool tpstats'
STDOUT:
Pool Name                    Active   Pending      Completed   Blocked  All time blocked
MutationStage                     0         0             26         0                 0
ViewMutationStage                 0         0              0         0                 0
ReadStage                         0         0              1         0                 0
RequestResponseStage              0         0              3         0                 0
ReadRepairStage                   0         0              1         0                 0
CounterMutationStage              0         0              0         0                 0
MiscStage                         0         0              0         0                 0
CompactionExecutor                0         0            314         0                 0
MemtableReclaimMemory             0         0             31         0                 0
PendingRangeCalculator            0         0              4         0                 0
GossipStage                       0         0           2876         0                 0
SecondaryIndexManagement          0         0              0         0                 0
HintsDispatcher                   0         0              0         0                 0
MigrationStage                    0         0              4         0                 0
MemtablePostFlush                 0         0             52         0                 0
ValidationExecutor                0         0              0         0                 0
Sampler                           0         0              0         0                 0
MemtableFlushWriter               0         0             31         0                 0
InternalResponseStage             0         0              4         0                 0
AntiEntropyStage                  0         0              0         0                 0
CacheCleanupExecutor              0         0              0         0                 0
Native-Transport-Requests         0         0              0         0                 0

Message type           Dropped
READ                         0
RANGE_SLICE                  0
_TRACE                       0
HINT                         0
MUTATION                     0
COUNTER_MUTATION             0
BATCH_STORE                  0
BATCH_REMOVE                 0
REQUEST_RESPONSE             0
PAGED_RANGE                  0
READ_REPAIR                  0
Writing file cassandra_cassandra_2018052539/cassandra_nodetool_tpstats_node-2-server__c7a678e7-dbcf-4fab-a9e1-97667a57fcd7.txt
(CLI) dcos task exec node-1-server__9a04a418-3e46-429b-aafe-9003ab1bf357 bash -c 'export JAVA_HOME=$(ls -d ${MESOS_SANDBOX}/jdk*/jre/) && export TASK_IP=$(${MESOS_SANDBOX}/bootstrap --get-task-ip) && CASSANDRA_DIRECTORY=$(ls -d ${MESOS_SANDBOX}/apache-cassandra-*/) && ${CASSANDRA_DIRECTORY}/bin/nodetool tpstats'
STDOUT:
Pool Name                    Active   Pending      Completed   Blocked  All time blocked
MutationStage                     0         0             23         0                 0
ViewMutationStage                 0         0              0         0                 0
ReadStage                         0         0              4         0                 0
RequestResponseStage              0         0             10         0                 0
ReadRepairStage                   0         0              2         0                 0
CounterMutationStage              0         0              0         0                 0
MiscStage                         0         0              0         0                 0
CompactionExecutor                0         0            384         0                 0
MemtableReclaimMemory             0         0             39         0                 0
PendingRangeCalculator            0         0              4         0                 0
GossipStage                       0         0           3307         0                 0
SecondaryIndexManagement          0         0              0         0                 0
HintsDispatcher                   0         0              0         0                 0
MigrationStage                    0         0              9         0                 0
MemtablePostFlush                 0         0             88         0                 0
ValidationExecutor                0         0              0         0                 0
Sampler                           0         0              0         0                 0
MemtableFlushWriter               0         0             39         0                 0
InternalResponseStage             0         0              4         0                 0
AntiEntropyStage                  0         0              0         0                 0
CacheCleanupExecutor              0         0              0         0                 0
Native-Transport-Requests         0         0              0         0                 0

Message type           Dropped
READ                         0
RANGE_SLICE                  0
_TRACE                       0
HINT                         0
MUTATION                     0
COUNTER_MUTATION             0
BATCH_STORE                  0
BATCH_REMOVE                 0
REQUEST_RESPONSE             0
PAGED_RANGE                  0
READ_REPAIR                  0
Writing file cassandra_cassandra_2018052539/cassandra_nodetool_tpstats_node-1-server__9a04a418-3e46-429b-aafe-9003ab1bf357.txt
(CLI) dcos task exec node-0-server__c98911be-c2ae-4121-aa78-60a6840bd886 bash -c 'export JAVA_HOME=$(ls -d ${MESOS_SANDBOX}/jdk*/jre/) && export TASK_IP=$(${MESOS_SANDBOX}/bootstrap --get-task-ip) && CASSANDRA_DIRECTORY=$(ls -d ${MESOS_SANDBOX}/apache-cassandra-*/) && ${CASSANDRA_DIRECTORY}/bin/nodetool tpstats'
STDOUT:
Pool Name                    Active   Pending      Completed   Blocked  All time blocked
MutationStage                     0         0             19         0                 0
ViewMutationStage                 0         0              0         0                 0
ReadStage                         0         0             53         0                 0
RequestResponseStage              0         0              2         0                 0
ReadRepairStage                   0         0              0         0                 0
CounterMutationStage              0         0              0         0                 0
MiscStage                         0         0              0         0                 0
CompactionExecutor                0         0            411         0                 0
MemtableReclaimMemory             0         0             32         0                 0
PendingRangeCalculator            0         0              4         0                 0
GossipStage                       0         0           3648         0                 0
SecondaryIndexManagement          0         0              0         0                 0
HintsDispatcher                   0         0              0         0                 0
MigrationStage                    0         0             14         0                 0
MemtablePostFlush                 0         0             80         0                 0
ValidationExecutor                0         0              0         0                 0
Sampler                           0         0              0         0                 0
MemtableFlushWriter               0         0             32         0                 0
InternalResponseStage             0         0              0         0                 0
AntiEntropyStage                  0         0              0         0                 0
CacheCleanupExecutor              0         0              0         0                 0
Native-Transport-Requests         0         0            152         0                 0

Message type           Dropped
READ                         0
RANGE_SLICE                  0
_TRACE                       0
HINT                         0
MUTATION                     0
COUNTER_MUTATION             0
BATCH_STORE                  0
BATCH_REMOVE                 0
REQUEST_RESPONSE             0
PAGED_RANGE                  0
READ_REPAIR                  0
Writing file cassandra_cassandra_2018052539/cassandra_nodetool_tpstats_node-0-server__c98911be-c2ae-4121-aa78-60a6840bd886.txt
```

```
$ ls cassandra_cassandra_2018052539
cassandra_nodetool_status_node-0-server__c98911be-c2ae-4121-aa78-60a6840bd886.txt
cassandra_nodetool_status_node-1-server__9a04a418-3e46-429b-aafe-9003ab1bf357.txt
cassandra_nodetool_status_node-2-server__c7a678e7-dbcf-4fab-a9e1-97667a57fcd7.txt
cassandra_nodetool_tpstats_node-0-server__c98911be-c2ae-4121-aa78-60a6840bd886.txt
cassandra_nodetool_tpstats_node-1-server__9a04a418-3e46-429b-aafe-9003ab1bf357.txt
cassandra_nodetool_tpstats_node-2-server__c7a678e7-dbcf-4fab-a9e1-97667a57fcd7.txt
dcos_marathon_task_list.json
dcos_task.json
service_configuration.json
service_plan_status_backup-azure.json
service_plan_status_backup-s3.json
service_plan_status_cleanup.json
service_plan_status_deploy.json
service_plan_status_recovery.json
service_plan_status_repair.json
service_plan_status_replace.json
service_plan_status_restore-azure.json
service_plan_status_restore-s3.json
service_pod_status.json
```

elezar commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,213 @@
+
+
+def main(argv):
+    if len(argv) < 2:
```
Could we use `argparse` instead?

mpereira replied:
Sure.

mpereira replied:
Done.

mpereira commented:
@elezar @porridge @nickbp I jush pushed what I belive is a usable first iteration of this tool. There's still a few quick things I want to do to consider this PR mergeable, but I believe the review process can start. I updated the PR description, PTAL.

porridge commented:
Before I dive into any of the code, some comments on usability regarding your mention:

```
Provide command so that people can run something like (maybe provide a wrapper script?):

docker run --rm mpereira/dcos-commons:diagnostics bash -c 'PYTHONPATH=/dcos-commons-dist/testing python3 /dcos-commons-dist/tools/create_service_bundle.py hdfs /data-services/hdfs'
```

We need to remember that the users are going to run this script in a hurry and under stress, when their service is in trouble. They also need to be able to run it in a variety of configurations, including air-gapped scenarios with weird proxy-only environments. Therefore running this script *must* be **insanely easy and robust**. We do not want them to spend time on debugging weird corner cases in using *this* script.

Therefore I think:
- docker is a no-go - will be difficult to use in air-gapped environments, and there will be problems with mounting output directory, etc.
- we could consider [building a `.par` file](https://github.com/google/subpar) but that also has its unique issues
- I think the most transparent option would be to build a plain-old tarball with the script and its dependencies. This way the users can easily inspect the contents before running the script.

The usage instructions should be:
1. fetch the tarball
1. unpack it with `tar -zxvf sdk-diagnostics-0.1.tgz`
1. (optional) look around, see, no trojans here
1. run the script: `cd sdk-diagnostics-0.1 && ./create_service_bundle.py`
1. pack the output: `tar -zcvf bundle.tgz bundle` (this step must be automated so it contains cluster UUID and timestamp in the name, otherwise there will be confusion about which bundle is which)
1. send it to us

To make this work seamlessly, `create_service_bundle.py` needs:
1. a `#!/usr/bin/env python3` shebang line
1. to prepend dependencies to `sys.path` at runtime, to avoid having to set `PYTHONPATH` manually

porridge commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+
+def browse_executor_tasks(agent_id: str, executor_sandbox_path: str) -> List[dict]:
+    executor_sandbox = browse_executor_sandbox(agent_id, executor_sandbox_path)
+    tasks_directory = next(
```
It's not obvious what this does. Returns the second occurrence of `.../tasks`, or a `None`?

elezar replied:
It seems as if one could just call `browse_executor_sandbox` with the `/tasks` path appended instead?

mpereira replied:
This has been fixed.

porridge commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+    def for_each_running_task_with_prefix(self, prefix, fn):
+        task_ids = list(
+            map(
+                lambda task: task["id"],
```
Use `itemgetter("id")` instead?

porridge replied:
Although the following would be much more readable IMHO:
```
task_ids = [task["id"] for task in self.running_tasks() if task["name"].startswith(prefix)]
```

elezar replied:
This is also more pythonic, and it may just be inexperience, but I find reading streaming code in Python always requires me to spend a couple of times reading it.

mpereira replied:
This has been fixed.

porridge commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+    package_name = argv[1]
+    service_name = argv[2]
+
+    def directory_date_string():
```
I'd prefer if these functions were moved to top level, if they do not need access to `main's` data.

mpereira replied:
Done.

porridge commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+import sdk_utils
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO, format="%(message)s")
```
Please include a timestamp here, it can be very valuable at times.

mpereira replied:
Timestamp is available with the latest changes.

elezar commented:
> I think the most transparent option would be to build a plain-old tarball with the script and its dependencies. This way the users can easily inspect the contents before running the script.

What about a standalone Golang binary? This also means that the users cannot edit the script locally.

elezar commented:
@mpereira quick question, have we looked at hooking into `dcos-diagnostics` at some level? This would reduce what we request from customers to a single thing.

elezar commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+
+
+def main(argv):
+    if len(argv) < 2:
```
Not critical now, but could we use `argparse`? Then we would get `usage` for free.

mpereira replied:
Done.

elezar commented on testing/sdk_utils.py:
```diff
@@ -169,3 +170,47 @@ def sort(coll):
+
+
+# https://github.com/pytoolz/toolz/blob/2bd9139d0d0e17d3426cb467b5f58b1fb6d8a439/toolz/itertoolz.py#L791
+def getter(index):
```
Why not just depend on `pytoolz/toolz` directly?  Also, I don't know about having these functions here in `sdk_utils`. The module is already a little too general, but at least it's *mostly* limited to DC/OS specifics.

mpereira replied:
I created a JIRA under the "technical tasks" epic for this: https://jira.mesosphere.com/browse/DCOS-42000

elezar commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+def _get_task_page(
+    offset: int, limit: int, framework_id: str = None, task_id: str = None
+) -> dict:
+    path = "/mesos/tasks?offset={}&limit={}".format(offset, limit)
```
Would it not make sense to rather download `state.json` once at some point and operate on that. Otherwise we may have changing state while performing these requests.

mpereira replied:
Yes! The script gets tasks through the cluster state with the latest changes.

elezar commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+        )
+
+
+def get_tasks(framework_id: str = None, task_id: str = None) -> dict:
```
Maybe align with what @nickbp is doing here: https://github.com/mesosphere/dcos-commons/blob/move-off-shakedown/testing/sdk_tasks.py#L135

mpereira replied:
The script gets tasks through the cluster state with the latest changes.

elezar commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+    page_tasks = page.get("tasks", [])
+    tasks_tasks = tasks.get("tasks", [])
+    merged_tasks = tasks_tasks + page_tasks
+    tasks["tasks"] = merged_tasks
```
Python is pass-by-reference by default. This means that this will change the `tasks` object in the calling function. Since we pass this to the `_get_tasks` function again, it should be OK though.

mpereira replied:
The script gets tasks through the cluster state with the latest changes.

elezar commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+    wait_fixed=DEFAULT_RETRY_WAIT, stop_max_attempt_number=DEFAULT_RETRY_MAX_ATTEMPTS
+)
+def get_framework_id(package_name: str, service_name: str) -> str:
+    framework_ids = sdk_cmd.svc_cli(
```
Is there an indication as to which CLI version supports this command?

mpereira replied:
The script gets Framework ID through the cluster state with the latest changes.

elezar commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+@retrying.retry(
+    wait_fixed=DEFAULT_RETRY_WAIT, stop_max_attempt_number=DEFAULT_RETRY_MAX_ATTEMPTS
+)
+def debug_agent_files(agent_id: str) -> List[str]:
```
Are these already part of the `dcos-diagnostics` bundle?

mpereira replied:
I'm not sure what you mean. This function is used to list the files on an agent so that the script can match the executor path. The files are not downloaded though.

elezar commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+    wait_fixed=DEFAULT_RETRY_WAIT, stop_max_attempt_number=DEFAULT_RETRY_MAX_ATTEMPTS
+)
+def download_agent_path(
+    agent_id: str, agent_file_path: str, output_file_path: str, chunk_size: int = 8192
```
Does it make sense to set chunk size as a parameter?

mpereira replied:
It already is a parameter?

elezar commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+            agent_id, executor_sandbox, output_task_directory, patterns_to_download
+        )
+
+        output_executor_directory = os.path.join(base_path, task_id, "executor")
```
Does this mean we will duplicate the executor logs for tasks that share an executor? Would it not make sense to more closely map the folder structure on the cluster itself?

mpereira replied:
> Does this mean we will duplicate the executor logs for tasks that share an executor?

Yes. This was intentional on my part. When I'm debugging a customer issue I'm generally looking at logs from a task-centric POV, so it made more sense to group them in this way, even though it is possible that some files are going to be duplicated. I'm going to add a TODO item for having the executor ID appended to the executor directory name, so that we have that information obvious just from looking at the file name.

elezar commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+        raise NotImplementedError
+
+
+class CassandraBundle(BaseTechBundle):
```
This is where things get really interesting. Up to here, I think everything could be done as part of `dcos-diagnostics` or the CLI in general -- but the service-specifics may need some more work.

elezar commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+    def task_exec(self, task_id, cmd):
+        full_cmd = " ".join(
+            [
+                "export JAVA_HOME=$(ls -d ${MESOS_SANDBOX}/jdk*/jre/) &&",
```
If we are targeting 1.9 cluster with this too, this isn't going to work as the task environment is not set there.

mpereira replied:
I added a limitation for now: 0d58aacee0f901f43bc8e17eafa9c2bcc14c08eb

elezar commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+        wait_fixed=DEFAULT_RETRY_WAIT,
+        stop_max_attempt_number=DEFAULT_RETRY_MAX_ATTEMPTS,
+    )
+    def task_exec(self, task_id, cmd):
```
Seems like this lends itself well to a different pattern where the `cmd` is augmented and passed to a the `task_exec` method of the base-class.

elezar commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+        return sdk_cmd.marathon_task_exec(task_id, "bash -c '{}'".format(full_cmd))
+
+    def create_stats_file(self, task_id):
+        command = "curl -s ${MESOS_CONTAINER_IP}:${PORT_HTTP}/_stats"
```
`MESOS_CONTAINER_IP` is only for post-1.9 clusters.

mpereira replied:
https://github.com/mesosphere/dcos-commons/commit/0d58aacee0f901f43bc8e17eafa9c2bcc14c08eb

elezar commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+
+
+BASE_TECH_BUNDLE = {
+    "beta-cassandra": CassandraBundle,
```
Could we just strip the `beta-` from the name before performing the lookup?

mpereira replied:
We could. I found it clearer to have a `package_name => bundle` mapping.

elezar commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+        stop_max_attempt_number=DEFAULT_RETRY_MAX_ATTEMPTS,
+    )
+    def create_task_file(self):
+        output = sdk_cmd.run_cli("task --json", print_output=False)
```
Note: This has an implicit 100 task limit.

elezar replied:
(also, would this not be available in `state.json`?)

mpereira replied:
The latest changes use Mesos state to get task information.

elezar commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+        output = sdk_cmd.svc_cli(
+            self.package_name,
+            self.service_name,
+            "pod status --json",
```
Does this show the status of all pods? What about `pod info` or is that just an alias?

mpereira replied:
Yes. `pod info` only works for a particular pod and doesn't output the full pod state (non-JSON output).

elezar commented on testing/sdk_diag.py:
```diff
@@ -277,13 +277,18 @@ def _dump_task_logs_for_agent(item: pytest.Item, agent_id: str, agent_tasks: lis
 
 
-def _dump_task_logs_for_task(item: pytest.Item, agent_id: str, agent_executor_paths: dict, task_entry: _TaskEntry):
+def _dump_task_logs_for_task(
```
This was already addressed in another PR, correct?

mpereira replied:
Yes. The latest diff doesn't show this fwiw since it has the latest master merged in.

surdy commented:
Just here to say love the effort here. This really needs to be a thing for debuggability perspective. I will give this a spin and see If I have any feedback from Support perspective.
BTW I was mocking up something similar on my side (when I was not aware of this effort). If anyone's interested : https://github.com/surdy/dcos-scripts/blob/master/generate_service_diagnostics

Its just a bash script wrapper around DC/OS CLI

elezar commented on testing/sdk_cmd.py:
```diff
@@ -79,7 +81,8 @@ def cluster_request(
     # consistently include slash prefix for clearer logging below
     cluster_path = "/" + cluster_path.lstrip("/")
-    log.info("(HTTP {}) {}".format(method.upper(), cluster_path))
+    if log_output:
```
An alternative would be to have a the global `log` object be a wrapper for the current `log` object and then enable or disable the logging in that class instead of having to add it to every line here?

elezar commented on testing/sdk_cmd.py:
```diff
@@ -69,6 +70,7 @@ def cluster_request(
     :param retry: Whether to retry the request automatically if an HTTP error (>=400) is returned.
     :param raise_on_error: Whether to raise a `requests.exceptions.HTTPError` if the response code is >=400.
                            Disabling this effectively implies `retry=False` where HTTP status is concerned.
+    :param log_output: Whether to log messages at all. Can be disabled to reduce noise.
```
An alternative is to use a different log level (`DEBUG` for the extra things?)

mpereira replied:
A `log_response` parameter was added on master: https://github.com/mesosphere/dcos-commons/blob/master/testing/sdk_cmd.py#L74

I just merged in the latest master and used that instead: https://github.com/mesosphere/dcos-commons/pull/2511/commits/e2e9221c687817a7283d610232b2141ae8320aed

mpereira commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+
+@retrying.retry(
+    wait_fixed=DEFAULT_RETRY_WAIT, stop_max_attempt_number=DEFAULT_RETRY_MAX_ATTEMPTS
+)
```
This was removed with the latest changes.

mpereira commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+        print_output=False,
+    )
+
+    assert len(framework_ids) == 1, "More than 1 Framework ID returned: {}".format(
```
The script gets Framework ID through the cluster state with the latest changes.

mpereira commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+
+class HdfsBundle(BaseTechBundle):
+    def create(self):
+        logger.info("Creating HDFS bundle")
```
I added a "noop" in the log lines for base techs that still don't do anything.

mpereira commented:
Since the last push I have made a few improvements:

-   [X] Modularized code
-   [X] Added a README with usage and development instructions
-   [X] Implemented more thorough error-handling
-   [X] Download Mesos state once and operate on that instead of querying Mesos API multiple times
-   [X] Added timestamps to log outputs
-   [X] Don't depend on CLI for `framework_id`
-   [X] The logger also creates a file in the bundle directory with the script's log output (in addition to stdout)
-   [X] Add interactive inputs and helpful output to make sure user is running diagnostics against the right cluster/service
-   [X] Provide single script that customers/support can download and run
-   [X] Write script usage documention and use argparse

And more that I'm probably missing

There are a bunch of things that can be improved and added, but I think save for obvious bugs we should release at least a first version of this and iterate.

mpereira commented on tools/create_service_bundle.py:
```diff
@@ -0,0 +1,568 @@
+        self.create_pod_status_file()
+        self.create_plans_status_files()
+        self.create_log_files()
+        BASE_TECH_BUNDLE[self.package_name](
```
Done.

mpereira commented:
> Just here to say love the effort here. This really needs to be a thing for debuggability perspective. I will give this a spin and see If I have any feedback from Support perspective.

@surdy absolutely! I added a README if you wanna try it out: https://github.com/mesosphere/dcos-commons/blob/dcos-37660-sdk-service-diagnostics/tools/diagnostics/README.md

mpereira commented:
Ping @elezar @porridge.

porridge commented on tools/diagnostics/agent.py:
```diff
@@ -0,0 +1,150 @@
+    return browse_executor_sandbox(agent_id, os.path.join(executor_sandbox_path, "tasks"))
+
+
+@retrying.retry(wait_fixed=DEFAULT_RETRY_WAIT, stop_max_attempt_number=DEFAULT_RETRY_MAX_ATTEMPTS)
```
`browse_task_sandbox` -> `browse_executor_tasks` -> `browse_executor_sandbox` -> `browse_agent_path`
Both the first and last function in this chain have `retrying.retry`.

Let me re-iterate my comment from July: https://github.com/mesosphere/dcos-commons/pull/2511#pullrequestreview-141811827

mpereira replied:
Good catch, thanks! 453c929a4ece907683fcfd672132c1e5aceac324

porridge commented on tools/diagnostics/agent.py:
```diff
@@ -0,0 +1,150 @@
+        raise_on_error=False,
+        log_response=False,
+    )
+
```
Can you drop these empty lines?

mpereira replied:
I find that vertical whitespace sometimes helps with readability, this case being an example. Since this was formatted with `black` does it make sense that we attempt to make it always remove all vertical whitespace inside functions? By default it preserves whatever it's there.

porridge replied:
I also agree whitespace sometimes helps with readability, but mostly in longer functions, so I wouldn't necessarily make black remove it unconditionally...
I just didn't feel like it was helping in this particular case. But if you feel stronger about this then feel free to leave them in.

mpereira replied:
Ok.

porridge commented on tools/diagnostics/base_tech_bundle/cassandra_bundle.py:
```diff
@@ -0,0 +1,49 @@
+
+        return sdk_cmd.marathon_task_exec(task_id, "bash -c '{}'".format(full_cmd))
+
+    def create_nodetool_status_file(self, task_id):
```
I had pointed this out already in https://github.com/mesosphere/dcos-commons/pull/2511#discussion_r206405469 - can you name these `download_` or `fetch_` or `retrieve_` rather than `create_`?
Creating a bundle sounds OK, but creating a status file sounds like it's our status we write, while in fact we just retrieve data from elsewhere.

mpereira replied:
For `create_log_files` I agree that `download_` or `fetch_` makes more sense, but here the code isn't downloading/fetching a nodetool status file, it's creating a file based on a command output, `nodetool status`.

ffb0d9dca1c4492fa9ff4f1f8f31ff66403bfbd5

porridge commented on tools/diagnostics/base_tech_bundle/cassandra_bundle.py:
```diff
@@ -0,0 +1,49 @@
+        return sdk_cmd.marathon_task_exec(task_id, "bash -c '{}'".format(full_cmd))
+
+    def create_nodetool_status_file(self, task_id):
+        rc, stdout, stderr = self.task_exec(task_id, "${CASSANDRA_DIRECTORY}/bin/nodetool status")
```
I had asked for this already in July: please do not hide stderr when something fails: https://github.com/mesosphere/dcos-commons/pull/2511#discussion_r206410196

mpereira replied:
f532a4d23db964b11b059bc4bb23340916577860

porridge commented on tools/diagnostics/bundle.py:
```diff
@@ -0,0 +1,21 @@
+log = logging.getLogger(__name__)
+
+
+class Bundle(object):
```
Do you need the `(object)` in Python3?

mpereira replied:
7e8bf60cded2dc95da43e0adc5ff19d9a11f2d43

porridge commented on tools/diagnostics/full_bundle.py:
```diff
@@ -0,0 +1,169 @@
+
+
+def directory_date_string() -> str:
+    return date.strftime(datetime.now(), "%Y%m%d%H%M%S")
```
Please make this UTC and use an ISO 8601 format.

mpereira replied:
b6947d737c960703a85fa48b98552a6f7cb5ac84

porridge commented on tools/diagnostics/full_bundle.py:
```diff
@@ -0,0 +1,169 @@
+        active_services = [s for s in services if is_service_active(s)]
+
+        if not active_services:
+            log.error("Could not find service named '%s'", self.service_name)
```
s/service/active service/

mpereira replied:
853098418f2be64f73532abb1ced3490bfece0ab

porridge commented on tools/diagnostics/service_bundle.py:
```diff
@@ -0,0 +1,153 @@
+
+log = logging.getLogger(__name__)
+
+DEFAULT_RETRY_WAIT = 1000
```
Same as in the other file.

mpereira replied:
3001569652caf2fb45657f1673bded4ec1556b4d

porridge commented on tools/diagnostics/service_bundle.py:
```diff
@@ -0,0 +1,153 @@
+
+
+class ServiceBundle(Bundle):
+    DOWNLOAD_FILES_WITH_PATTERNS = ["^stdout(\.\d+)?$", "^stderr(\.\d+)?$"]
```
I believe these are called something else in sdk 0.50+, please check and amend.

mpereira replied:
Looking at the task sandbox for Kubernetes tasks (which run on SDK 0.50+) it looks like the log files are still called `stdout.N` and `stderr.N`. Do you remember where you saw the thing that gave you that impression?

porridge replied:
@nickbp said so here: https://jira.mesosphere.com/browse/DCOS-40654?focusedCommentId=171361&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-171361

porridge replied:
Looks like it's in master only so far? Still, worth preparing for that.

mpereira replied:
33f4e940b80cc8226c494ccaca996aa2fe593a70

porridge commented on tools/create_service_diagnostics_bundle.py:
```diff
@@ -0,0 +1,225 @@
+    if rc != 0:
+        (success, cluster_name_or_error) = current_cluster_name()
+
+        if "dcos auth login" or "Missing required config parameter" in stderr:
```
This does not do what you think :-)
```
>>> if "bug" or "impossible" in "badger":
...   print("peekaboo!")
... 
peekaboo!
>>> 
```

mpereira replied:
Good catch caaa42f3661c6650175921179b0a0e26bbd2aa0d

porridge commented on tools/create_service_diagnostics_bundle.py:
```diff
@@ -0,0 +1,225 @@
+
+
+def main(argv):
+    (rc, should_proceed, args) = preflight_check()
```
You don't need the parentheses.

mpereira replied:
ad4836c093791a9d53f728b08c23a3b0e2f03778

porridge commented on tools/create_service_diagnostics_bundle.py:
```diff
@@ -0,0 +1,225 @@
+        if answer.strip().lower() in ["n", "no", "false"]:
+            return 0
+
+    (rc, _) = FullBundle(
```
Ditto.

mpereira replied:
ad4836c093791a9d53f728b08c23a3b0e2f03778

mpereira commented on testing/sdk_utils.py:
```diff
@@ -205,6 +205,61 @@ def get_in(keys, coll, default=None):
         return default
 
 
+def sort(coll):
```
55f7c9d6972f2fbbd4561c48695b5b09c7e6cedf

mpereira commented on testing/sdk_utils.py:
```diff
@@ -205,6 +205,61 @@ def get_in(keys, coll, default=None):
+    return coll
+
+
+def invert_dict(d: dict) -> dict:
```
55f7c9d6972f2fbbd4561c48695b5b09c7e6cedf

mpereira commented on tools/diagnostics/README.md:
````diff
@@ -0,0 +1,109 @@
+   readonly SCRIPT_PATH="/dcos-commons/tools/create_service_diagnostics_bundle.py"
+   ```
+
+   *TODO*: instead of doing all this manually make the script take a flag and do that automatically (e.g.
````
That is a good idea.

mpereira replied:
I implemented it: 8cd1f54efc1991de09255c7278f0239523eaf9eb

Thanks for the idea, this is much better than passing parameters!

mpereira commented on tools/diagnostics/__init__.py:
```diff
@@ -0,0 +1,7 @@
+from diagnostics.service_bundle import ServiceBundle
+import diagnostics.agent as agent
+
+__all__ = ["FullBundle", "Bundle", "ServiceBundle", "BaseTechBundle", "agent"]
```
abcc9854870e348baf6fbc19832e37bf7f6ce057

mpereira commented on tools/diagnostics/agent.py:
```diff
@@ -0,0 +1,150 @@
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_RETRY_WAIT = 1000
```
3001569652caf2fb45657f1673bded4ec1556b4d

elezar commented on tools/diagnostics/base_tech_bundle/elastic_bundle.py:
```diff
@@ -30,7 +25,12 @@ def task_exec(self, task_id, cmd):
-        self.write_file("elasticsearch_stats_{}.json".format(task_id), stdout)
+        if rc != 0 or stderr:
+            logger.error(
+                "Could not get Elasticsearch /_stats\nstdout: '{}'\nstderr: '{}'", stdout, stderr
```
In python logging on uses `%s` and not `{}`.

mpereira replied:
Fixed: https://github.com/mesosphere/dcos-commons/pull/2511/commits/0bc6f8e624c73735825ba6c422a9566cfa5b0e87

mpereira commented on tools/diagnostics/README.md:
````diff
@@ -0,0 +1,109 @@
+1. Build a Docker image tagged with the desire version
+
+   ```bash
+   docker build -t "mpereira/dcos-commons:diagnostics-${VERSION}" .
````
I still need to figure out how to get push credentials to our Docker hub...

porridge replied:
Sure, but please add a TODO.

mpereira replied:
https://jira.mesosphere.com/browse/DCOS-42310

mpereira commented on tools/create_service_diagnostics_bundle.sh:
```diff
@@ -0,0 +1,50 @@
+
+readonly VERSION='v0.1.0'
+readonly DCOS_COMMONS_DIRECTORY="/dcos-commons-dist"
+readonly DOCKER_IMAGE="mpereira/dcos-commons:diagnostics-${VERSION}"
```
I still need to figure out how to get push credentials to our Docker hub. Perhaps we could follow up after the PR is merged on updating this? I'll look into it tomorrow.

porridge replied:
OK, but please add a TODO.

mpereira replied:
https://jira.mesosphere.com/browse/DCOS-42310

mpereira commented on testing/sdk_utils.py:
```diff
@@ -205,6 +205,61 @@ def get_in(keys, coll, default=None):
+    return dict((v, k) for k, v in d.items())
+
+
+# https://github.com/pytoolz/toolz/blob/2bd9139d0d0e17d3426cb467b5f58b1fb6d8a439/toolz/itertoolz.py#L791
```
I created this JIRA to address it: https://jira.mesosphere.com/browse/DCOS-42000

porridge replied:
I'm not sure our lawyers would be so relaxed about this. In fact by pushing this branch to a public directory you've violated the license already. :-/

mpereira replied:
I pushed a PR adding toolz as a dependency and merged it into this PR: https://github.com/mesosphere/dcos-commons/pull/2511/commits/f30073504483e8e5022faa25ad25b9afc4e08fb4

mpereira commented on tools/create_service_diagnostics_bundle.sh:
```diff
@@ -0,0 +1,50 @@
+
+for requirement in ${REQUIREMENTS}; do
+  if ! [[ -x $(command -v "${requirement}") ]]; then
+    echo "You need to install '${requirement}' to run this script"
```
I weighted the short-term implementation alternatives (detailed in the design document) and found that Docker was the most appealing, even with its shortcomings.

The long-term implementation should be what @elezar recommended:
> What about a `dcos` subcommand that targets services? Remember that most of our services (except Kafka and Spark at the moment) use the same CLI. We could have a "service-diag" CLI command that is installed and can target any service:
> dcos package install --yes --cli service-diagnostics
> dcos service-diagnostics collect --name=/my/elastic/service
>
> Existing examples of such CLIs are:
> * The enterprise CLI
> * The tunnel CLI

With this implementation we just want to provide customers something that is better than the current situation, which they can always fall-back on in case of problems running this script.

porridge replied:
It might be most appealing to us in the short term, but I'm afraid we'll just have to add a tarball form of distribution soon, once we hit a customer who cannot use docker.

AFAICT the only difference between using a tarball and the docker image found in https://docs.google.com/document/d/1STxZPbV5OJ8D-jj1u72jKLlTKq3FsG2nniER3_qWNr8/edit#heading=h.g95cohdwgsdd is that "the environment might be unknown" which seems very vague to me. OTOH a customer not being to use docker is a very concrete limitation.

I'm afraid we'll have to fix the "Python environment (dependencies, versions, ENV, etc.) management is very bad" case either way - whether we do it in a tarballed directory with  venv+pip or in a docker image with system-level pip.

Or am I missing something?

mpereira replied:
> I'm afraid we'll have to fix the "Python environment (dependencies, versions, ENV, etc.) management is very bad" case either way - whether we do it in a tarballed directory with venv+pip or in a docker image with system-level pip.

That's true. In the Docker image that's already "fixed" so there's nothing to do :)

What about this: we release this using a Docker image and if some customer (and by "customer" here I'm assuming operations engineers or developers, the people who reach out to support and which I'd expect to have Docker installed in their system) can't use Docker *then* we work on figuring out how to distribute it with a tarball containing the whole controlled environment?

porridge replied:
We're actually far from "nothing to do" even in the docker case. Not even all direct dependencies are pinned to a fixed version, never mind indirect dependencies. We had a CI outage not so long ago exactly because of this problem.

So whenever we rebuild the docker image we will pick up new versions of some potentially broken dependencies - I had filed DCOS-41913 about this.

I also only now realized that you are reusing the `mesosphere/dcos-commons` docker image (which we use for CI) for this script. Additionally the script you added has no integration tests, so nobody will even notice if they break the docker image for the purpose of running this script, as long as it keeps working for CI.

Do you think I could convince you to add a simple integration test that runs this script against any of our services just to make sure it does not crash and produces *something*?

porridge replied:
> Right. Since the script runs off a versioned image (e.g. mpereira/dcos-commons:diagnostics-v0.1.0 (and not off a mutable tag like latest), unless the tag is force pushed the script environment will remain unchanged for its version.

That's very good, but I had a different scenario in mind: once we fix even a typo in the diagnostic script, we need to rebuild the docker image, which means we will pick up any changes that happened to any unpinned dependency we use.

elezar commented on tools/diagnostics/create_service_diagnostics_bundle.py:
```diff
@@ -0,0 +1,224 @@
+
+    if args.get("should_prompt_user"):
+        answer = input("\nProceed? [Y/n]: ")
+        if answer.strip().lower() in ["n", "no", "false"]:
```
Why not `answer.strip().lower() not in ["yes", y"]:` instead?

mpereira replied:
Because the default action is `Y`, indicated by its capitalization. Pressing enter with no actual input should proceed, for example. I guess this can be `if answer.strip().lower() not in ["yes", "y", ""]:`?

elezar replied:
The problem is that `"foo"`, `"nooooooo"` etc. will also be taken as `yes` --> I would rather check for the empty string explicitly then.

mpereira replied:
f76163cf3d6f7c7e8000e5652520583cc5417aec

elezar commented on tools/diagnostics/create_service_diagnostics_bundle.sh:
```diff
@@ -0,0 +1,97 @@
+readonly DCOS_COMMONS_SCRIPT_PATH="tools/diagnostics"
+
+function is_dcos_commons_repository_script() {
+  [[ ${SCRIPT_DIRECTORY} = *${DCOS_COMMONS_SCRIPT_PATH} ]]
```
Should one check for a git repo with a valid remote too?

mpereira replied:
I originally had that but removed it since this could be run from an expanded tarball directory too, for example.

elezar commented on tools/diagnostics/create_service_diagnostics_bundle.sh:
```diff
@@ -0,0 +1,97 @@
+docker run \
+       -it \
+       -v "$(pwd)/${BUNDLES_DIRECTORY}:${CONTAINER_BUNDLES_DIRECTORY}" \
+       -v "${HOME}/.dcos":/root/.dcos \
```
Should this be mounted `:ro`?

mpereira replied:
The script install service CLIs (e.g. `dcos package install elastic --cli --yes`) which mutate `/root/.dcos` in the container. Do you think it's fine that this isn't RO or should we think about maybe not mounting but just copying the host `.dcos` to the container `.dcos`?

porridge replied:
Alternatively we could make it fail if required CLI plugin is not installed, and have the user do it for us with consent.

elezar commented on tools/diagnostics/create_service_diagnostics_bundle.sh:
```diff
@@ -0,0 +1,97 @@
+}
+
+function usage () {
+  echo 'Usage: ./create_service_diagnostics_bundle.sh <package name> <service name> [<proceed>]'
```
Would it make sense to just pass the arguments directly to the python script -- then the command line will be the same? This will also allow proceed to work as expected.

mpereira replied:
Good idea: d47214d860e09f212ffd3c0f8d6d9a61aece4fef

elezar commented on tools/diagnostics/full_bundle.py:
```diff
@@ -0,0 +1,181 @@
+
+        return directory_name
+
+    def create(self) -> (int, "FullBundle"):
```
Is this how one specifies a class return?

mpereira replied:
[Yup](https://www.python.org/dev/peps/pep-0484/#forward-references).

mockingbird2 replied:
Couldn't we use it without the quotation marks here?

mpereira replied:
No, because `FullBundle` still isn't defined at this point. The link above has more detail.

elezar commented on tools/diagnostics/service_bundle.py:
```diff
@@ -0,0 +1,148 @@
+
+
+class ServiceBundle(Bundle):
+    DOWNLOAD_FILES_WITH_PATTERNS = ["^stdout(\.\d+)?(\.log)?$", "^stderr(\.\d+)?(\.log)?$"]
```
I don't believe the `.log` extensions are required.

mpereira replied:
01f49021bf51b8c54d7779985e2dc1f776abb341

porridge commented on tools/diagnostics/service_bundle.py:
```diff
@@ -14,7 +14,7 @@
 
 class ServiceBundle(Bundle):
-    DOWNLOAD_FILES_WITH_PATTERNS = ["^stdout(\.\d+)?$", "^stderr(\.\d+)?$"]
+    DOWNLOAD_FILES_WITH_PATTERNS = ["^stdout(\.\d+)?(\.log)?$", "^stderr(\.\d+)?(\.log)?$"]
```
Yeah, sorry for the confusion, looks like I had misunderstood Nick.

mpereira replied:
01f49021bf51b8c54d7779985e2dc1f776abb341

porridge commented:
Looks good, modulo the docker distribution method and the licensing issue.

mpereira commented on tools/diagnostics/agent.py:
```diff
@@ -0,0 +1,147 @@
+        log_response=False,
+    )
+
+    if is_http_server_error(response.status_code):
```
The intention here is to only raise an exception (trigger a retry, effectively) on server errors. It doesn't make sense to retry client errors IMO. Server errors might be transient, requests resulting in client errors will yield the same error responses.

mpereira commented on tools/diagnostics/agent.py:
```diff
@@ -0,0 +1,147 @@
+        log_response=False,
+    )
+
+    if is_http_server_error(response.status_code):
```
Same reason as above.

mpereira commented on tools/diagnostics/README.md:
```diff
@@ -0,0 +1,120 @@
+     aws s3 cp \
+       --acl=public-read \
+       tools/diagnostics/create_service_diagnostics_bundle.sh \
+       "s3://infinity-artifacts/dcos-commons/diagnostics/${VERSION}/create_service_diagnostics_bundle.sh"
```
> Does this mean that users need to download the script AND the Docker image?
No. The README instructions are complete--users only need to download the shell script, which when run will fetch the Docker image through `docker run`.

mpereira commented on tools/diagnostics/README.md:
```diff
@@ -0,0 +1,120 @@
+
+Will create bundle for:
+  Package:         cassandra
+  Package version: 2.3.0-3.0.16-rc3
```
c075745c4faaac4ae6aaa6d5d207c7c896a76dd3

mpereira commented on tools/diagnostics/create_service_diagnostics_bundle.py:
```diff
@@ -0,0 +1,224 @@
+
+
+def main(argv):
+    rc, should_proceed, args = preflight_check()
```
Good point 86cc9dfa6972d1633d373d62d48bcd433e95b60d

mpereira commented on tools/diagnostics/create_service_diagnostics_bundle.py:
```diff
@@ -0,0 +1,224 @@
+
+# Dependencies:
+# - DC/OS CLI
+# - jq
```
Historical reasons maybe 00ece2450ca5511a38f8949af114dcb119fe10e0

mpereira commented on tools/diagnostics/create_service_diagnostics_bundle.py:
```diff
@@ -0,0 +1,224 @@
+
+# Notes:
+# - When run in the dcos-commons docker image the CLI version is specified in the
+#   test_requirements.txt file.
```
Good point d6ee67b212ab0bcdc83a201f77ca37adf6533100

mpereira commented on tools/diagnostics/base_tech_bundle/elastic_bundle.py:
```diff
@@ -0,0 +1,40 @@
+    def task_exec(self, task_id, cmd):
+        full_cmd = " ".join(
+            [
+                "export JAVA_HOME=$(ls -d ${MESOS_SANDBOX}/jdk*/jre/) &&",
```
Makes sense. I'll leave that as a future opportunity for improvement.

porridge commented:
I think the only remaining reservation from myself are the `mpereira` prefixes on docker image tags, and the lack of a regression test.

There are also a handful unaddressed comments from @elezar such as [this one](https://github.com/mesosphere/dcos-commons/pull/2511/files#r218759710).

mpereira commented on tools/diagnostics/README.md:
````diff
@@ -0,0 +1,120 @@
+   readonly VERSION='vx.y.z'
+   ```
+
+1. Build a Docker image tagged with the desired version
````
Yep https://jira.mesosphere.com/browse/DCOS-42310

mpereira commented:
@porridge @elezar I believe I've addressed all feedback, PTAL. I also just pushed a change to make the script also fetch the scheduler offers: https://github.com/mesosphere/dcos-commons/pull/2511/commits/ee4bc7beb047c5a5cf67f5ed79640d28d748d2c6

porridge commented:
Yeah, but it's not a silver bullet:
```
porridge@beczulka:~$ mkdir funny\ name
porridge@beczulka:~$ cat heh.sh
bash -c "ls $*"
porridge@beczulka:~$ bash heh.sh funny\ name
ls: cannot access 'funny': No such file or directory
ls: cannot access 'name': No such file or directory
porridge@beczulka:~$ 
```

elezar commented on testing/sdk_hosts.py:
```diff
@@ -35,6 +36,19 @@ def vip_host(service_name, vip_name, port=-1):
     return _to_host(_safe_name(vip_name), _safe_name(service_name), VIP_HOST_SUFFIX, port)
 
 
+def scheduler_vip_host(service_name, vip_name, port=-1):
```
Nit: The `vip_name` comes first in the resultant `vip`. Not a blocker.

mpereira replied:
I didn't like it too much too but I decided to make it consistent with `vip_host`.

elezar commented on tools/diagnostics/create_service_diagnostics_bundle.sh:
```diff
@@ -0,0 +1,91 @@
+done
+
+readonly BUNDLES_DIRECTORY="service-diagnostic-bundles"
+readonly DOCKER_IMAGE="mpereira/dcos-commons:diagnostics-${VERSION}"
```
We're still using your personal `docker` image.

mpereira replied:
This is gonna be addressed in https://jira.mesosphere.com/browse/DCOS-42310

elezar replied:
I think we should do it now in the following steps:
1. Change the image to `mesosphere/dcos-commons:diagnostics-${VERSION}`
1. Build the docker image off this branch
1. Merge!
(This would also address the ticket you mention above)

In the long run we would need to think a bit about what it meant to "release" this script.

elezar replied:
But I'm OK to merge this if you feel that we should handle all docker-related issues in the linked JIRA ticket.

mpereira commented:
> The problem is that "foo", "nooooooo" etc. will also be taken as yes --> I would rather check for the empty string explicitly then.

f76163cf3d6f7c7e8000e5652520583cc5417aec

mpereira commented:
I tested the script manually with this and saw no issues:
```bash
#!/usr/bin/env bash

set -eu -o pipefail

readonly USERNAME=dcos-dev
readonly PASSWORD=dogfoodinit

readonly CLUSTERS=(
  "soak110"
  "soak111s"
  "soak112s"
)

readonly SERVICES=(
  "cassandra           /data-services/cassandra"
  "confluent-kafka     /data-services/confluent-kafka"
  "confluent-zookeeper /data-services/confluent-zookeeper"
  "datastax-dse        /data-services/datastax-dse"
  "datastax-ops        /data-services/datastax-ops"
  "elastic             /data-services/elastic"
  "hdfs                /data-services/hdfs"
  "kafka-zookeeper     /data-services/kafka-zookeeper"
)

for cluster in "${CLUSTERS[@]}"; do
  dcos cluster setup \
       --insecure \
       --username="${USERNAME}" \
       --password="${PASSWORD}" \
       "https://${cluster}.testing.mesosphe.re/" || true

  for service in "${SERVICES[@]}"; do
    package_name="$(echo "${service}" | awk '{print $1}')"
    service_name="$(echo "${service}" | awk '{print $2}')"

    ./tools/diagnostics/create_service_diagnostics_bundle.sh \
      --package-name="${package_name}" \
      --service-name="${service_name}" \
      --yes
  done
done
```
